### ISSUE TITLE
Add Kafka Streams runner

### ISSUE THREAD

**Issue by kennknowles** [apache/beam#18479]
Kafka Streams ([https://kafka.apache.org/documentation/streams](https://kafka.apache.org/documentation/streams)) has more and more features that could make it a viable candidate for a streaming runner. It uses DataFlow-like model.

 

Please look at the [Design Document](https://docs.google.com/document/d/1mNqERvvV8oGI_O4tGewH2Kgkq6PQGv3ylmxnaTRBqH8/edit?usp=sharing) and add comments.

Imported from Jira [BEAM-2466](https://issues.apache.org/jira/browse/BEAM-2466). Original Jira may contain additional context.
Reported by: klorand.

**Comment by wilsonwang371** [comment]
I am actually very interested in this topic. Generally, it is possible to have beam running on a generic faas system that with MQ support?

**Comment by je-ik** [comment]
There are several requirements that must be met. Essentially:
 a) preserving per-partition order of records (i.e. records emitted in order from one distributed producer must not overtake each other when consumed)
 b) producer must be able to enqueue output records for a specific consumer (e.g. assigning a key of a output record, all records with same key must then be consumed by the same instance of downstream consumer)
 c) producer must be able to send record to all downstream consumers (i.e. producer must know how many consumers there - possibly - is)
 d) there must be some kind of support of state commit, either at the end of bundle, during bundle commit (dataflow model), or as a flowing checkpoint barrier (flink model), there must be a way to safely store state in a distributed fault tolerant storage and be able to possibly restore the complete state from that committed state
 
Having these conditions met I think it should be possible (though quite hard) to implement Beam runner on top of it. Kafka definitely has all four (even without Kafka streams).

**Comment by junaiddshaukat** [comment]
Hi, I'm interested in working on the portable Kafka Streams Runner 
as part of GSoC 2026 (if it got selected).

I've been contributing to the Apache Beam TypeScript SDK — my merged 
PRs include LRU cache eviction for CachingStateProvider (#37214) and 
CombinePerKeyPrecombineOperator (#37466). I'm now ramping up on the 
Beam portability framework and Kafka Streams to prepare for this project.

I've read through the design document and the previous attempts by 
Kai Jiang and Kyle Winkelman. A few questions:

1. Should the GSoC skeleton target the portable runner API (Fn API / 
   gRPC), or a classic translation-based approach?
2. What's the minimum set of transforms the skeleton should support — 
   Read, ParDo, GroupByKey?
3. Is there a preference for using the Kafka Streams DSL vs the 
   lower-level Processor API for the translation layer?

I'll be studying the Flink portable runner as a reference implementation. 
Happy to discuss the design further.

**Comment by je-ik** [comment]
Hi @junaiddshaukat, great to hear about your interest. It's also cool that you have such a big insight into the Beam ecosystem. Maybe we can put down the design document for the GSoC together? Generally speaking, my point of view would be that:
 1) the skeleton should target the FnAPI to have full support of all SDKs (it should be possible to optimize the translation for Java SDK later)
 2) the skeleton should be "useful", i.e. it should be possible to run at least some basic Pipelines, that would imply we should target to implement at least:
  - Read
  - stateless ParDo
  - GBK
  - CBK
  - Window
  - optional: stateful ParDo, later splittable DoFn
 3) this is crucial question that needs deeper analysis to make sure that the DSL aligns correctly with the Apache Beam model. My intuition here would be that the Processor API would be a better choice, because it is flexible enough to support the model and yet it should not force us to manually manage state, etc.

We would also need to define the design of watermarks, bundles (we must pay attention to Beam's guarantees and make the bundling compatible), timers (if we would support stateful ParDo).

Would you like to try to sketch a design document we could iterate on? When we create some basic doc, we can then share it on the dev@ list to get more feedback and ensure we don't hit a wall during implementation.

**Comment by junaiddshaukat** [comment]
Thank you @je-ik for the detailed response and the clear direction! 
I'd love to sketch a design document and iterate on it with you.

Your points give a solid foundation to start from:

1. **Fn API** as the target — agreed, this ensures all SDKs work from 
   day one with room for Java-specific optimizations later.
2. **Processor API** over DSL — makes sense given the flexibility 
   needed for Beam's model. I'll analyze the alignment more deeply 
   in the design doc.
3. **Minimum transforms**: Read, stateless ParDo, GBK, CBK, Window — 
   with stateful ParDo and splittable DoFn as stretch goals.

I'll start drafting the design document this week, covering:
- High-level architecture (pipeline translation flow)
- Mapping of Beam primitives to Kafka Streams Processor API constructs
- Watermark propagation strategy
- Bundle management and Beam's guarantees
- Timer support considerations

I'll share the initial draft here for your feedback before we take it 
to the dev@ list. Looking forward to collaborating!

**Comment by junaiddshaukat** [comment]
Hi @je-ik, I've put together an initial draft of the design document 
for the portable Kafka Streams runner:

https://docs.google.com/document/d/1BBMURhSG4SxPcvvnKMTrmnKCr_jhXL6R4TBDBW7zsy8/edit?tab=t.0

It covers the high-level architecture, pipeline translation approach 
(following the Flink portable runner pattern), transform mappings for 
Read, ParDo, GBK, CBK, Window, and Flatten, along with sections on 
watermark management, bundle handling, and state management.

I've included a few architecture diagrams and listed open questions at 
the end — particularly around watermark derivation from Kafka consumer 
positions, repartitioning strategy for GBK, and bundle-commit alignment 
for exactly-once semantics.

Looking forward to your feedback so we can iterate on this before 
sharing it on the dev@ list.

**Comment by je-ik** [comment]
That's awesome document, I will walk through it ASAP, during the weekend latest. Thanks a lot for this!

**Comment by junaiddshaukat** [comment]
Hi @je-ik,

I’ve revised the [Design Document](https://docs.google.com/document/d/1BBMURhSG4SxPcvvnKMTrmnKCr_jhXL6R4TBDBW7zsy8/edit?usp=sharing) based on your suggestions. Could you please take a look when you have a moment and let me know your feedback?

Thank you!

**Comment by MansiSingh17** [comment]
Hi @je-ik  and @junaiddshaukat , I've been studying the design document for the portable Kafka Streams runner and wanted to share a few observations and questions.

I'm also interested in contributing to this project for GSoC 2026 — I've been contributing to the Beam TypeScript SDK (merged PRs #37214 and #37466) and have been ramping up on the portability framework.

- On the Impulse bootstrap topic (Section 4.1): The `__beam_impulse` topic approach makes sense for single-pipeline scenarios, but I'm wondering how multi-pipeline isolation works. If two pipelines run concurrently, do they share the same topic and state store? Could this cause interference between pipelines, or is the applicationId sufficient to isolate state per pipeline?

- On watermark advancement without data (Section 12, Open Question 3): This is listed as an open question but I didn't see a proposed direction. The Flink runner uses a dedicated watermark emit thread for this — is that a viable approach here, or does Kafka Streams' single-threaded processor model make that problematic?

- On error handling for ExecutableStage: I didn't see anything in the design about what happens when the SDK harness fails to process a record. Is the expectation that Kafka's retry mechanism handles this, or would we need explicit dead letter queue handling similar to other runners?

- On the ValidatesRunner test scope (Section 10): The doc mentions running "a subset" of ValidatesRunner tests, would it make sense to define the specific test classes that should pass as part of the GSoC deliverable? That would make the acceptance criteria clearer.

Happy to dig into any of these further or help with the design doc if useful.

**Comment by junaiddshaukat** [comment]
> Hi [@je-ik](https://github.com/je-ik) and [@junaiddshaukat](https://github.com/junaiddshaukat) , I've been studying the design document for the portable Kafka Streams runner and wanted to share a few observations and questions.
> 
> I'm also interested in contributing to this project for GSoC 2026 — I've been contributing to the Beam TypeScript SDK (merged PRs [#37214](https://github.com/apache/beam/pull/37214) and [#37466](https://github.com/apache/beam/pull/37466)) and have been ramping up on the portability framework.
> 
> * On the Impulse bootstrap topic (Section 4.1): The `__beam_impulse` topic approach makes sense for single-pipeline scenarios, but I'm wondering how multi-pipeline isolation works. If two pipelines run concurrently, do they share the same topic and state store? Could this cause interference between pipelines, or is the applicationId sufficient to isolate state per pipeline?
> * On watermark advancement without data (Section 12, Open Question 3): This is listed as an open question but I didn't see a proposed direction. The Flink runner uses a dedicated watermark emit thread for this — is that a viable approach here, or does Kafka Streams' single-threaded processor model make that problematic?
> * On error handling for ExecutableStage: I didn't see anything in the design about what happens when the SDK harness fails to process a record. Is the expectation that Kafka's retry mechanism handles this, or would we need explicit dead letter queue handling similar to other runners?
> * On the ValidatesRunner test scope (Section 10): The doc mentions running "a subset" of ValidatesRunner tests, would it make sense to define the specific test classes that should pass as part of the GSoC deliverable? That would make the acceptance criteria clearer.
> 
> Happy to dig into any of these further or help with the design doc if useful.


Hi @MansiSingh17, thanks for the interest in the project and the 
thoughtful questions.

Just a small correction — PRs #37214 and #37466 are my contributions 
to the Beam TypeScript SDK. Perhaps you meant to reference different 
PR numbers?

Regarding your questions:

1. **Impulse topic isolation:** Good point. The `applicationId` in 
   Kafka Streams provides namespace isolation — each application has 
   its own consumer group and state stores. For multi-pipeline 
   scenarios, each pipeline would use a unique `applicationId`, so 
   they wouldn't share state or interfere with each other. The 
   bootstrap topic name could also include the `applicationId` as 
   a prefix for additional isolation.

2. **Watermark advancement without data:** The Source's reader has a 
   `getCurrentTimestamp()` method that provides timestamps independently 
   of input data. We periodically call this and update the output 
   watermark as `min(timestamp_of_all_unfinished_readers_in_task)`. 
   This was discussed with @je-ik in the design doc review.

3. **Error handling for ExecutableStage:** Good question — this is 
   something we can address during implementation. The Flink runner's 
   `ExecutableStageDoFnOperator` handles SDK harness failures by 
   failing the bundle and retrying. We'd follow a similar pattern.

4. **ValidatesRunner test scope:** Agreed that defining specific test 
   classes would help. We can refine this during implementation based 
   on which transforms are supported at each stage.

The design document has been through several review rounds with @je-ik.

**Comment by MansiSingh17** [comment]
@junaiddshaukat Hi @junaiddshaukat, thanks for the correction — I apologize for the mix-up. My contributions are actually to the Beam Python SDK: #37672 (fix GroupBy snippet tests and re-enable skipped assertions) and #37674 (make GCS filesystem lookup lazy to match S3 behavior). I shouldn't have referenced TypeScript SDK PRs in my original comment.
Thank you for the detailed answers. A couple of follow-up thoughts:

- On watermark advancement: For partitions that are idle with no new records arriving, would we still need a periodic wall-clock punctuation to trigger the watermark check? Or does the source reader advance independently of data flow?

- On error handling: Since we have manual commit control, failed bundles could potentially replay from the last committed offset rather than failing the whole job. Worth thinking about early given it affects how bundle boundaries are designed.

 I'm happy to contribute to any part of this project. Is there an aspect of the implementation where an additional contributor would be helpful?

**Comment by je-ik** [comment]
> I'm happy to contribute to any part of this project. Is there an aspect of the implementation where an additional contributor would be helpful?

This is great idea! Yes, I think if we have more contributors, we can slightly change the scope and add more features, namely full stateful ParDo and splittable DoFn support. I'm not sure what the exact rules of GSoC are, can a single task have multiple people working on it? Maybe @damccorm might know?

**Comment by junaiddshaukat** [comment]
> > I'm happy to contribute to any part of this project. Is there an aspect of the implementation where an additional contributor would be helpful?
> 
> This is great idea! Yes, I think if we have more contributors, we can slightly change the scope and add more features, namely full stateful ParDo and splittable DoFn support. I'm not sure what the exact rules of GSoC are, can a single task have multiple people working on it? Maybe [@damccorm](https://github.com/damccorm) might know?

As far as I know, each GSoC project can have only one contributor. But if GSoC allows it would be great

**Comment by junaiddshaukat** [comment]
> [@junaiddshaukat](https://github.com/junaiddshaukat) Hi [@junaiddshaukat](https://github.com/junaiddshaukat), thanks for the correction — I apologize for the mix-up. My contributions are actually to the Beam Python SDK: [#37672](https://github.com/apache/beam/pull/37672) (fix GroupBy snippet tests and re-enable skipped assertions) and [#37674](https://github.com/apache/beam/pull/37674) (make GCS filesystem lookup lazy to match S3 behavior). I shouldn't have referenced TypeScript SDK PRs in my original comment. Thank you for the detailed answers. A couple of follow-up thoughts:
> 
> * On watermark advancement: For partitions that are idle with no new records arriving, would we still need a periodic wall-clock punctuation to trigger the watermark check? Or does the source reader advance independently of data flow?
> * On error handling: Since we have manual commit control, failed bundles could potentially replay from the last committed offset rather than failing the whole job. Worth thinking about early given it affects how bundle boundaries are designed.
> 
> I'm happy to contribute to any part of this project. Is there an aspect of the implementation where an additional contributor would be helpful?

On the follow-up questions:

**Watermark with idle partitions:** Yes, we'd need a periodic 
wall-clock punctuation to check source readers even when no data 
arrives. The source reader's `getCurrentTimestamp()` provides the 
watermark independently of data flow, but we need to poll it 
periodically. KS `punctuate(WALL_CLOCK_TIME)` can handle this — 
it runs independently of stream-time advancement.

**Error handling / bundle replay:** Good point. With exactly-once 
and manual commit, a failed bundle means the transaction is aborted 
and offsets aren't committed. KS will replay from the last committed 
offset on recovery, effectively retrying the failed bundle. This 
aligns with Beam's bundle retry semantics. We should document this 
in the design doc.

Is anything else you want to know, feel free to ask here.

**Comment by damccorm** [comment]
> > > I'm happy to contribute to any part of this project. Is there an aspect of the implementation where an additional contributor would be helpful?
> > 
> > 
> > This is great idea! Yes, I think if we have more contributors, we can slightly change the scope and add more features, namely full stateful ParDo and splittable DoFn support. I'm not sure what the exact rules of GSoC are, can a single task have multiple people working on it? Maybe [@damccorm](https://github.com/damccorm) might know?
> 
> As far as I know, each GSoC project can have only one contributor. But if GSoC allows it would be great

I think this is correct, there can only be one contributor per project. I think one of the goals is to decouple the fates of the 2 contributors (e.g. it avoids the problem if one contributor does a great job and the other does not).

With that said, if we can do some scoping and divide this up, though, there's not a reason we couldn't have 2 GSOC projects which collectively add a more fully featured runner. We'd just want to be clear on a division of responsibilities as much as is possible, and formally they would be 2 projects, not 1. @je-ik if you want to try this, I think it is fine

**Comment by je-ik** [comment]
@damccorm Thanks for clarification. I'm not sure if this GH issue is the right place to discuss this or should we swtich to the ASF Jira for thiat? That problem is - it is not clear how to split this into two independent projects, if one builds on the other. We have "create a skeleton to support basic Pipelines" and we can do "support more of the Beam model to support broader range of Pipelines".  The first project has to (at least partially) exist before the other can start.

**Comment by MansiSingh17** [comment]
Hi @je-ik  and @damccorm , I'd love to take on the second project scope if this split moves forward.

On the dependency concern — I think we can handle it by defining Project 2 to start with the skeleton from Project 1 as a baseline assumption, and in the first few weeks coordinate closely with @junaiddshaukat to build on top of the foundation as it gets established. The stateful ParDo and splittable DoFn work can be largely designed and planned during that time even before all the skeleton pieces land.

A rough split could be:
- Project 1: Job Server, Pipeline Translator, Read, stateless ParDo, GBK, Combine, Window, Flatten, basic ValidatesRunner tests
- Project 2: Stateful ParDo, splittable DoFn, timer support, extended ValidatesRunner coverage

Happy to discuss further here or move to the mailing list if that's more appropriate. I can also start drafting a proposal outline for the second scope if that helps clarify feasibility.

**Comment by MansiSingh17** [comment]
Hi @je-ik and @damccorm,  wanted to follow up on the two-project split discussion. I can start drafting a proposal outline for Project 2 (stateful ParDo, splittable DoFn, timer support, extended ValidatesRunner coverage) so we have something concrete to react to. Does that make sense as a next step?

**Comment by je-ik** [comment]
Hi @MansiSingh17, I was thinking about this and essentially yes, it might be possible to split the projects, probably something like:
 a) design common data objects (probably based on protobuffs) that will be used for communication between transforms (e.g. watermarks, windwedvalues, runtime metadata)
 b) design (possibly abstract or interface) watermark manager that will be used in all transforms that need to compute input/output watermarks
 c) create a common skeleton of pipeline translator and translation context

This might be enough to split these projects, but Project B will _still_ be heavily dependent on Project A. Without Project A being successfully merged, Project B will not be able to succeed. So this is a risk you would probably have to face.

**Comment by MansiSingh17** [comment]
Hi @je-ik , thanks for thinking this through — the split you've outlined makes sense to me.

Having the shared data objects (protobuf-based windowed values, watermarks, runtime metadata), the watermark manager interface, and the pipeline translator skeleton as the boundary between the two projects gives Project 2 a clear foundation to build on, and makes the interface contract explicit enough that design work can start in parallel.

I'm fully aware of the dependency risk you mentioned — if Project 1 doesn't merge, Project 2 can't succeed. My plan would be to spend the first few weeks of the program in close coordination with @junaiddshaukat, designing and prototyping the stateful ParDo and splittable DoFn layers against the interfaces as they take shape, so there's no dead time waiting on merges.

I'll start drafting a Project 2 proposal outline this week covering the scope, milestones, and how I'd handle the dependency. Happy to share it here for your feedback before I submit.

**Comment by junaiddshaukat** [comment]
.take-issue
